Powershell for Management Groups

rajshah11 opened the pull request:
<!-- DO NOT DELETE THIS TEMPLATE -->

## Description
<!--
Please add an informative description that covers the changes made by the pull request.

If applicable, reference the bug/issue that this pull request fixes here.
-->
This PR includes cmdlets and tests for Management Groups.
---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md).**
- [x] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/preview/documentation/help-generation.md).**
- [x] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log).**

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/preview/documentation/cleaning-up-commits.md).
- [x] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.
- [x] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [x] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://github.com/Azure/azure-powershell/wiki/PowerShell-Cmdlet-Design-Guidelines#supportsshouldprocess).
- [x] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThru` parameter.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [x] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [x] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [x] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

cormacpayne commented on src/ResourceManager/ManagementGroups/Commands.ManagementGroups/Cmdlets/UpdateAzureRmManagementGroup.cs:
```diff
@@ -0,0 +1,50 @@
+
+        [Parameter(ParameterSetName = Constants.ParameterSetNames.GroupOperationsParameterSet, Mandatory = false,
+            HelpMessage = Constants.HelpMessages.Force, Position = 3)]
+        public SwitchParameter Force { get; set; }
```
@rajshah11 same comment about whether or not we need the `Force` parameter

rajshah11 replied:
Removed.

cormacpayne commented on src/ResourceManager/ManagementGroups/Commands.ManagementGroups/Cmdlets/UpdateAzureRmManagementGroup.cs:
```diff
@@ -0,0 +1,50 @@
+        public string GroupName { get; set; }
+
+        [Parameter(ParameterSetName = Constants.ParameterSetNames.GroupOperationsParameterSet, Mandatory = false,
+            HelpMessage = Constants.HelpMessages.DisplayName, Position = 1)]
```
@rajshah11 same comment about removing the position attribute from optional parameters in this cmdlet

rajshah11 replied:
Done. Thanks!

cormacpayne commented on src/ResourceManager/ManagementGroups/Commands.ManagementGroups/Cmdlets/UpdateAzureRmManagementGroup.cs:
```diff
@@ -0,0 +1,50 @@
+
+namespace Microsoft.Azure.Commands.ManagementGroups.Cmdlets
+{
+    [Cmdlet("Update", "AzureRmManagementGroup", DefaultParameterSetName = Constants.ParameterSetNames.GroupOperationsParameterSet, SupportsShouldProcess = true, ConfirmImpact = ConfirmImpact.Medium), OutputType(typeof(string))]
```
@rajshah11 same comment about updating the output type of this cmdlet to reflect what is written to the output stream below

rajshah11 replied:
Done. Thanks!

cormacpayne commented on src/ResourceManager/ManagementGroups/Commands.ManagementGroups/Commands.ManagementGroups.csproj:
```diff
@@ -0,0 +1,197 @@
+    <WarningLevel>4</WarningLevel>
+  </PropertyGroup>
+  <ItemGroup>
+    <Reference Include="Hyak.Common, Version=1.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
```
@rajshah11 please import the `Common.Dependencies.targets` file, [like so](https://github.com/Azure/azure-powershell/blob/preview/src/ResourceManager/Billing/Commands.Billing/Commands.Billing.csproj#L92). This will allow you to remove almost all of the packages you are referencing in this file.

rajshah11 replied:
Done. Thanks!!

cormacpayne commented on src/ResourceManager/ManagementGroups/Commands.ManagementGroups/Commands.ManagementGroups.csproj:
```diff
@@ -0,0 +1,197 @@
+      <Link>AzureRM.ManagementGroups.psd1</Link>
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
+    <None Include="app.config" />
```
@rajshah11 please remove this reference to `app.config`

rajshah11 replied:
Removed.

cormacpayne commented on src/ResourceManager/ManagementGroups/Commands.ManagementGroups/Commands.ManagementGroups.csproj:
```diff
@@ -0,0 +1,197 @@
+  </ItemGroup>
+  <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
+  <Import Project="..\packages\Microsoft.Bcl.Build.1.0.14\tools\Microsoft.Bcl.Build.targets" Condition="Exists('..\packages\Microsoft.Bcl.Build.1.0.14\tools\Microsoft.Bcl.Build.targets')" />
+  <Target Name="EnsureBclBuildImported" BeforeTargets="BeforeBuild" Condition="'$(BclBuildImported)' == ''">
```
@rajshah11 you can remove this Bcl target

rajshah11 replied:
Removed.

cormacpayne commented on src/ResourceManager/ManagementGroups/Commands.ManagementGroups/Properties/AssemblyInfo.cs:
```diff
@@ -0,0 +1,35 @@
+// You can specify all the values or you can default the Build and Revision Numbers 
+// by using the '*' as shown below:
+// [assembly: AssemblyVersion("1.0.*")]
+[assembly: AssemblyVersion("1.0.0.0")]
```
@rajshah11 same comment about keeping this values in line with the module version

rajshah11 replied:
Done.

cormacpayne commented on src/ResourceManager/ManagementGroups/Commands.ManagementGroups/packages.config:
```diff
@@ -0,0 +1,18 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<packages>
+  <package id="Hyak.Common" version="1.0.2" targetFramework="net452" />
```
@rajshah11 similarly to your `.csproj` file, you can remove all of the common packages (it should be everything but the `Microsoft.Azure.Management.ResourceManager` package)

rajshah11 replied:
Done.

cormacpayne commented:
@rajshah11 Hey Raj, the build is currently failing because it cannot find version 1.7.0-preview of `Microsoft.Azure.Management.ResourceManager`. You can do one of  two things:

(1) If your changes are merged into the .NET SDK repository and the package is ready to be published, you can publish it to NuGet

(2) Create a signed NuGet package that you can place in `tools/LocalFeed` and use for testing until the package is pushed to NuGet

rajshah11 commented on src/ResourceManager/ManagementGroups/Commands.ManagementGroups.Test/Properties/AssemblyInfo.cs:
```diff
@@ -0,0 +1,35 @@
+// You can specify all the values or you can default the Build and Revision Numbers 
+// by using the '*' as shown below:
+// [assembly: AssemblyVersion("1.0.*")]
+[assembly: AssemblyVersion("1.0.0.0")]
```
Done. Thanks!

rajshah11 commented on src/ResourceManager/ManagementGroups/Commands.ManagementGroups/Cmdlets/GetAzureRmManagementGroup.cs:
```diff
@@ -0,0 +1,61 @@
+    /// <summary>
+    /// Get-AzureRmManagementGroup Cmdlet
+    /// </summary>
+    [Cmdlet(VerbsCommon.Get,"AzureRmManagementGroup", DefaultParameterSetName = Constants.ParameterSetNames.GroupOperationsParameterSet, SupportsShouldProcess = true, ConfirmImpact = ConfirmImpact.Medium), OutputType(typeof(string))]
```
Done. Thanks!

rajshah11 commented on src/ResourceManager/ManagementGroups/Commands.ManagementGroups/Cmdlets/GetAzureRmManagementGroup.cs:
```diff
@@ -0,0 +1,61 @@
+    /// <summary>
+    /// Get-AzureRmManagementGroup Cmdlet
+    /// </summary>
+    [Cmdlet(VerbsCommon.Get,"AzureRmManagementGroup", DefaultParameterSetName = Constants.ParameterSetNames.GroupOperationsParameterSet, SupportsShouldProcess = true, ConfirmImpact = ConfirmImpact.Medium), OutputType(typeof(string))]
```
Done. Thanks!

rajshah11 commented on src/ResourceManager/ManagementGroups/Commands.ManagementGroups/Cmdlets/GetAzureRmManagementGroup.cs:
```diff
@@ -0,0 +1,61 @@
+                }
+                else
+                {
+                    var response = ManagementGroupsApiClient.ManagementGroups.List();
```
Done. Thanks!

rajshah11 commented on src/ResourceManager/ManagementGroups/Commands.ManagementGroups/Cmdlets/NewAzureRmManagementGroup.cs:
```diff
@@ -0,0 +1,52 @@
+    /// <summary>
+    /// New-AzureRmManagementGroup Cmdlet
+    /// </summary>
+    [Cmdlet(VerbsCommon.New, "AzureRmManagementGroup", DefaultParameterSetName = Constants.ParameterSetNames.GroupOperationsParameterSet, SupportsShouldProcess = true, ConfirmImpact = ConfirmImpact.Medium), OutputType(typeof(string))]
```
Done. Thanks!

rajshah11 commented on src/ResourceManager/ManagementGroups/Commands.ManagementGroups/Cmdlets/NewAzureRmManagementGroup.cs:
```diff
@@ -0,0 +1,52 @@
+    /// <summary>
+    /// New-AzureRmManagementGroup Cmdlet
+    /// </summary>
+    [Cmdlet(VerbsCommon.New, "AzureRmManagementGroup", DefaultParameterSetName = Constants.ParameterSetNames.GroupOperationsParameterSet, SupportsShouldProcess = true, ConfirmImpact = ConfirmImpact.Medium), OutputType(typeof(string))]
```
Done. Thanks!

rajshah11 commented on src/ResourceManager/ManagementGroups/Commands.ManagementGroups/Cmdlets/NewAzureRmManagementGroup.cs:
```diff
@@ -0,0 +1,52 @@
+        public string GroupName { get; set; }
+
+        [Parameter(ParameterSetName = Constants.ParameterSetNames.GroupOperationsParameterSet, Mandatory = false,
+            HelpMessage = Constants.HelpMessages.DisplayName, Position = 1)]
```
Done. Thanks!

rajshah11 commented on src/ResourceManager/ManagementGroups/Commands.ManagementGroups/Cmdlets/NewAzureRmManagementGroup.cs:
```diff
@@ -0,0 +1,52 @@
+
+        [Parameter(ParameterSetName = Constants.ParameterSetNames.GroupOperationsParameterSet, Mandatory = false,
+            HelpMessage = Constants.HelpMessages.Force, Position = 3)]
+        public SwitchParameter Force { get; set; }
```
Not necessary as such. Removing it.

rajshah11 commented on src/ResourceManager/ManagementGroups/Commands.ManagementGroups/Cmdlets/NewAzureRmManagementGroupSubscription.cs:
```diff
@@ -0,0 +1,48 @@
+    /// </summary>
+    [Cmdlet(VerbsCommon.New, "AzureRmManagementGroupSubscription",
+         DefaultParameterSetName = Constants.ParameterSetNames.GroupOperationsParameterSet,
+         SupportsShouldProcess = true, ConfirmImpact = ConfirmImpact.Medium), OutputType(typeof(string))]
```
Done. Thanks!

rajshah11 commented on src/ResourceManager/ManagementGroups/Commands.ManagementGroups/Cmdlets/NewAzureRmManagementGroupSubscription.cs:
```diff
@@ -0,0 +1,48 @@
+
+        [Parameter(ParameterSetName = Constants.ParameterSetNames.GroupOperationsParameterSet, Mandatory = false,
+            HelpMessage = Constants.HelpMessages.Force, Position = 2)]
+        public SwitchParameter Force { get; set; }
```
Yes. Not necessary. Thanks!

rajshah11 commented on src/ResourceManager/ManagementGroups/Commands.ManagementGroups/Cmdlets/NewAzureRmManagementGroupSubscription.cs:
```diff
@@ -0,0 +1,48 @@
+                        string.Format(Resource.NewManagementGroupSubShouldProcessTarget, SubscriptionId, GroupName),
+                        string.Format(Resource.NewManagementGroupSubShouldProcessAction, SubscriptionId, GroupName)))
+                {
+                    ManagementGroupsApiClient.ManagementGroupSubscriptions.Create(GroupName, SubscriptionId.ToString());
```
Yes. We are not returning anything. PassThru implemented. Thanks!

rajshah11 commented on src/ResourceManager/ManagementGroups/Commands.ManagementGroups/Cmdlets/RemoveAzureRmManagementGroup.cs:
```diff
@@ -0,0 +1,43 @@
+    /// </summary>
+    [Cmdlet(VerbsCommon.Remove, "AzureRmManagementGroup",
+         DefaultParameterSetName = Constants.ParameterSetNames.GroupOperationsParameterSet,
+         SupportsShouldProcess = true, ConfirmImpact = ConfirmImpact.Medium), OutputType(typeof(string))]
```
Done. Thanks!

rajshah11 commented on src/ResourceManager/ManagementGroups/Commands.ManagementGroups/Cmdlets/RemoveAzureRmManagementGroup.cs:
```diff
@@ -0,0 +1,43 @@
+
+        [Parameter(ParameterSetName = Constants.ParameterSetNames.GroupOperationsParameterSet, Mandatory = false,
+            HelpMessage = Constants.HelpMessages.Force, Position = 1)]
+        public SwitchParameter Force { get; set; }
```
Removed

rajshah11 commented on src/ResourceManager/ManagementGroups/Commands.ManagementGroups/Cmdlets/RemoveAzureRmManagementGroup.cs:
```diff
@@ -0,0 +1,43 @@
+                        string.Format(Resource.RemoveManagementGroupShouldProcessTarget, GroupName),
+                        string.Format(Resource.RemoveManagementGroupShouldProcessAction, GroupName)))
+                {
+                    ManagementGroupsApiClient.ManagementGroups.Delete(GroupName);
```
Implemented. Thanks!

rajshah11 commented on src/ResourceManager/ManagementGroups/Commands.ManagementGroups/Cmdlets/RemoveAzureRmManagementGroupSubscription.cs:
```diff
@@ -0,0 +1,48 @@
+    /// </summary>
+    [Cmdlet(VerbsCommon.Remove, "AzureRmManagementGroupSubscription",
+         DefaultParameterSetName = Constants.ParameterSetNames.GroupOperationsParameterSet,
+         SupportsShouldProcess = true, ConfirmImpact = ConfirmImpact.Medium), OutputType(typeof(string))]
```
Removed.

rajshah11 commented on src/ResourceManager/ManagementGroups/Commands.ManagementGroups/Cmdlets/RemoveAzureRmManagementGroupSubscription.cs:
```diff
@@ -0,0 +1,48 @@
+
+        [Parameter(ParameterSetName = Constants.ParameterSetNames.GroupOperationsParameterSet, Mandatory = false,
+            HelpMessage = Constants.HelpMessages.Force, Position = 2)]
+        public SwitchParameter Force { get; set; }
```
Removed.

rajshah11 commented on src/ResourceManager/ManagementGroups/Commands.ManagementGroups/Cmdlets/RemoveAzureRmManagementGroupSubscription.cs:
```diff
@@ -0,0 +1,48 @@
+                        string.Format(Resource.RemoveManagementGroupSubShouldProcessTarget, SubscriptionId,GroupName),
+                        string.Format(Resource.RemoveManagementGroupSubShouldProcessAction, SubscriptionId, GroupName)))
+                {
+                    ManagementGroupsApiClient.ManagementGroupSubscriptions.Delete(GroupName, SubscriptionId.ToString());
```
Done.

rajshah11 commented on src/ResourceManager/ManagementGroups/Commands.ManagementGroups/Cmdlets/UpdateAzureRmManagementGroup.cs:
```diff
@@ -0,0 +1,50 @@
+
+namespace Microsoft.Azure.Commands.ManagementGroups.Cmdlets
+{
+    [Cmdlet("Update", "AzureRmManagementGroup", DefaultParameterSetName = Constants.ParameterSetNames.GroupOperationsParameterSet, SupportsShouldProcess = true, ConfirmImpact = ConfirmImpact.Medium), OutputType(typeof(string))]
```
Done.

rajshah11 commented:
@cormacpayne I think my build is now failing because of the wrong wix file. However, I cannot generate the wix file because my local build is failing while generation of a help file. The error is 

```RM.EventHub\help\Get-AzureRmEventHubGeoDRConfiguration.md:25:(83) 'The 
    Expect Heading"
   At C:\Program Files\WindowsPowerShell\Modules\platyPS\0.8.1\platyPS.psm1:1372 char:9
   +         $maml = $t.NodeModelToMamlModel($model)
  +         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
       + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
       + FullyQualifiedErrorId : HelpSchemaException````

Is there a way to skip this error and proceed with the build?

cormacpayne commented:
@rajshah11 that error is coming from our `GenerateHelp.ps1` script, which uses the `platyPS` module to generate XML help from the markdown help in the repository. Would you mind providing the full error message you're seeing? The first line you copied was `Expect Heading"`, which should be the string that tells you the markdown file that is improperly formatted (if you grab the rest of the string before that line).

rajshah11 commented:
@cormacpayne 
```


      Directory: D:\GPC\azure-powershell\src\Package\Debug\ResourceManager\AzureResourceManager\AzureRM.EventGrid


  Mode                LastWriteTime         Length Name

  ----                -------------         ------ ----

  -a----         2/2/2018   2:25 PM         242382 Microsoft.Azure.Commands.EventGrid.dll-Help.xml

  ForEach-Object : Exception calling "NodeModelToMamlModel" with "1" argument(s): "D:\GPC\azure-powershell\src\Package\
  De
  bug\ResourceManager\AzureResourceManager\AzureRM.EventHub\help\Get-AzureRmEventHubGeoDRConfiguration.md:25:(83) 'The
  '
   Expect Heading"
  At C:\Users\shahraj\Documents\WindowsPowerShell\Modules\platyPS\0.8.3\platyPS.psm1:1463 char:22
  +     $markdownFiles | ForEach-Object {
  +                      ~~~~~~~~~~~~~~~~
      + CategoryInfo          : NotSpecified: (:) [ForEach-Object], MethodInvocationException
      + FullyQualifiedErrorId : HelpSchemaException,Microsoft.PowerShell.Commands.ForEachObjectCommand

D:\GPC\azure-powershell\build.proj(208,5): error MSB3073: The command ""C:\Windows\System32\WindowsPowerShell\v1.0\powe
rshell.exe" -NonInteractive -NoLogo -NoProfile -Command "$ProgressPreference = 'SilentlyContinue';. D:\GPC\azure-powers
hell\tools\GenerateHelp.ps1 -ValidateMarkdownHelp -GenerateMamlHelp -BuildConfig Debug "" exited with code 1.
Done Building Project "D:\GPC\azure-powershell\build.proj" (default targets) -- FAILED.


Build FAILED.
```

rajshah11 commented:
@cormacpayne Did you get a chance to look into this?

cormacpayne commented:
@rajshah11 Hey Raj, sorry for the delayed response. It looks like the markdown file for  [`Get-AzureRmEventHubGeoDRConfiguration`](https://github.com/Azure/azure-powershell/blob/preview/src/ResourceManager/EventHub/Commands.EventHub/help/Get-AzureRmEventHubGeoDRConfiguration.md) is missing a `Description` header, which is causing that to fail. For some reason though, it isn't failing on our CI. If you add a `Description` header to this file (you can see any other markdown files as an example), then the build should succeed for you. We'll also open a PR in the `preview` branch to fix this issue.

cormacpayne commented on src/ResourceManager/ManagementGroups/Commands.ManagementGroups/Cmdlets/UpdateAzureRmManagementGroup.cs:
```diff
@@ -0,0 +1,60 @@
+namespace Microsoft.Azure.Commands.ManagementGroups.Cmdlets
+{
+    [Cmdlet("Update", "AzureRmManagementGroup", DefaultParameterSetName = Constants.ParameterSetNames.GroupOperationsParameterSet, SupportsShouldProcess = true), OutputType(typeof(PSManagementGroupNoChildren))]
+    public class UpdateAzureRmManagementGroup : AzureManagementGroupsCmdletBase
```
@rajshah11 same comment about supporting the `InputObject` parameter set

rajshah11 replied:
Done. Thanks

cormacpayne commented on src/ResourceManager/ManagementGroups/Commands.ManagementGroups/Commands.ManagementGroups.csproj:
```diff
@@ -0,0 +1,123 @@
+  </ItemGroup>
+  <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
+  <Import Project="..\..\..\..\tools\Common.Dependencies.targets" />
+  <Import Project="..\packages\Microsoft.Bcl.Build.1.0.14\tools\Microsoft.Bcl.Build.targets" Condition="Exists('..\packages\Microsoft.Bcl.Build.1.0.14\tools\Microsoft.Bcl.Build.targets')" />
```
@rajshah11 you should be able to remove this import

rajshah11 replied:
Remove. Thanks

rajshah11 commented on src/ResourceManager/ManagementGroups/AzureRM.ManagementGroups.psd1:
```diff
@@ -0,0 +1,127 @@
+        # IconUri = ''
+
+        # ReleaseNotes of this module
+        ReleaseNotes = '* Add support for online help
```
Done. Thanks!

rajshah11 commented on src/ResourceManager/ManagementGroups/AzureRM.ManagementGroups.psd1:
```diff
@@ -0,0 +1,127 @@
+    - Run Get-Help with the -Online parameter to open the online help in your default Internet browser'
+
+        # External dependent modules of this module
+        # ExternalModuleDependencies = ''
```
Done. Thanks.

rajshah11 commented on src/ResourceManager/ManagementGroups/Commands.ManagementGroups/Models/PSManagementGroupNoChildren.cs:
```diff
@@ -0,0 +1,71 @@
+
+namespace Microsoft.Azure.Commands.ManagementGroups.Models
+{
+    public class PSManagementGroupNoChildren
```
The issue is with output formatting. Children are shown only when -Expand is supplied. When it is not, the output has Children with null (even though the group may have children). This might give an impression to the user that the group has no children and to avoid this behavior I added a new object which does not have children at all.

rajshah11 commented on src/ResourceManager/ManagementGroups/Commands.ManagementGroups/Cmdlets/NewAzureRmManagementGroupSubscription.cs:
```diff
@@ -0,0 +1,77 @@
+                    string.Format(Resource.NewManagementGroupSubShouldProcessTarget, SubscriptionId, GroupName),
+                    string.Format(Resource.NewManagementGroupSubShouldProcessAction, SubscriptionId, GroupName)))
+                {
+                    ManagementGroupsApiClient.ManagementGroupSubscriptions.Create(GroupName, SubscriptionId.ToString());
```
No

rajshah11 commented on src/ResourceManager/ManagementGroups/Commands.ManagementGroups/Cmdlets/RemoveAzureRmManagementGroupSubscription.cs:
```diff
@@ -0,0 +1,78 @@
+    [Cmdlet(VerbsCommon.Remove, "AzureRmManagementGroupSubscription",
+         DefaultParameterSetName = Constants.ParameterSetNames.GroupOperationsParameterSet,
+         SupportsShouldProcess = true), OutputType(typeof(bool))]
+    public class RemoveAzureRmManagementGroupSubscription : AzureManagementGroupAutoRegisterCmdletBase
```
We don't have a Get for subscription in a group. I don't think piping should be supported here. Correct me if I am wrong. Thanks

rajshah11 commented on src/ResourceManager/ManagementGroups/Commands.ManagementGroups/Cmdlets/NewAzureRmManagementGroupSubscription.cs:
```diff
@@ -0,0 +1,77 @@
+                {
+                    if (context.Subscription.Id != SubscriptionId.ToString())
+                    {
+                        Utility.AzureManagementGroupAutoRegisterSubscription(SubscriptionId.ToString(), context);
```
The way auto-registration is handled right now doesn't work for tenant level resources (It catches an exception which is not thrown for tenant level resources) This was discussed in the review meeting and it was decided to include the changes now while you work to change autoreg on your side

rajshah11 commented on src/ResourceManager/ManagementGroups/Commands.ManagementGroups/Cmdlets/RemoveAzureRmManagementGroupSubscription.cs:
```diff
@@ -0,0 +1,78 @@
+                {
+                    if (context.Subscription.Id != SubscriptionId.ToString())
+                    {
+                        Utility.AzureManagementGroupAutoRegisterSubscription(SubscriptionId.ToString(), context);
```
@cormacpayne Replied above

rajshah11 commented on src/ResourceManager/ManagementGroups/Commands.ManagementGroups/Cmdlets/GetAzureRmManagementGroup.cs:
```diff
@@ -0,0 +1,75 @@
+                    var response = ManagementGroupsApiClient.ManagementGroups.List();
+                    var items = response.Select(managementGroup => new PSManagementGroupInfo(managementGroup))
+                        .ToList();
+                    WriteObject(items);
```
Done. Thanks!

rajshah11 commented on src/ResourceManager/ManagementGroups/Commands.ManagementGroups/Cmdlets/RemoveAzureRmManagementGroup.cs:
```diff
@@ -0,0 +1,61 @@
+    [Cmdlet(VerbsCommon.Remove, "AzureRmManagementGroup",
+         DefaultParameterSetName = Constants.ParameterSetNames.GroupOperationsParameterSet,
+         SupportsShouldProcess = true), OutputType(typeof(bool))]
+    public class RemoveAzureRmManagementGroup : AzureManagementGroupsCmdletBase
```
Done. Thanks

rajshah11 commented on src/ResourceManager/ManagementGroups/Commands.ManagementGroups.Test/ScenarioTests/UpdateAzureRmManagementGroupTests.ps1:
```diff
@@ -0,0 +1,102 @@
+
+function Test-UpdateManagementGroup
+{
+    $response = Update-AzureRmManagementGroup -GroupName TestPSNewGroup3
```
Done. Thanks!

rajshah11 commented on src/ResourceManager/ManagementGroups/Commands.ManagementGroups.Test/ScenarioTests/RemoveAzureRmManagementGroupTests.ps1:
```diff
@@ -0,0 +1,26 @@
+
+function Test-RemoveManagementGroup
+{
+    $response = Remove-AzureRmManagementGroup -GroupName TestPSNewGroup3
```
Done. Thanks!

rajshah11 commented on src/ResourceManager/ManagementGroups/Commands.ManagementGroups.Test/ScenarioTests/GetAzureRmManagementGroupTests.ps1:
```diff
@@ -0,0 +1,141 @@
+
+function Test-ListManagementGroups
+{
+    $response = Get-AzureRmManagementGroup
```
Done. Thanks!

rajshah11 commented:
@cormacpayne made all the changes and the build is passing too. Just FYI we have moved out of the ResourceManager SDK for a SDK of our own. I have reverted the package references changes and made the new changes accordingly

maddieclayton commented:
Closing in favor of the new PR
